lib/site-inspector.rb

```diff
@@ -127,37 +134,213 @@ def to_json
+    details = {}
+
+    # Don't follow redirects for first ping.
+    response = request(ssl, www, false)
+
+
+    # For HTTPS: examine the full range of possibilities.
+    if ssl
+      if response.return_code == :ok
+        details[:https_valid] = true
+
+      # Bad certificate chain.
+      elsif response.return_code == :ssl_cacert
+        details[:https_valid] = false
+        details[:https_bad_chain] = true
+        response = request(ssl, www, false, false, true)
+        # Bad everything.
+        if response.return_code == :peer_failed_verification
+          details[:https_bad_name] = true
+          response = request(ssl, www, false, false, false)
+        end
+      # Bad hostname.
+      elsif response.return_code == :peer_failed_verification
+        details[:https_valid] = false
+        details[:https_bad_name] = true
+        response = request(ssl, www, false, true, false)
+        # Bad everything.
+        if response.return_code == :ssl_cacert
+          details[:https_bad_chain] = true
+          response = request(ssl, www, false, false, false)
+        end
+      end
+    end
+
+    # If we ended up with a failure, return it.
+    details[:status] = response.response_code
+    return details if response.response_code == 0
+
+    headers = Hash[response.headers.map{ |k,v| [k.downcase,v] }]
+    # details[:headers] = headers
+
+
+    # HSTS only takes effect when delivered over valid HTTPS.
+    details[:hsts] = !!(ssl and details[:https_valid] and headers["strict-transport-security"])
+    details[:hsts_header] = headers["strict-transport-security"]
+
+
+    # If it's a redirect, go find the ultimate response starting from this combo.
+    redirect_code = response.response_code.to_s.start_with?("3")
+    location_header = headers["location"]
+    if redirect_code and location_header
+      details[:redirect] = true
+
+      ultimate_response = request(ssl, www, true, !details[:https_bad_chain], !details[:https_bad_name])
+      uri_original = URI(ultimate_response.request.url)
+      uri_eventual = URI(ultimate_response.effective_url)
+
+      details[:redirect_to] = uri_eventual.to_s
+      details[:redirect_to_external] = ((uri_original.hostname != uri_eventual.hostname) or (uri_original.scheme != uri_eventual.scheme))
+      details[:redirect_to_https] = (uri_eventual.scheme == "https")
+
+      details[:live] = ultimate_response.success?
+
+    # otherwise, judge it here
+    else
+      details[:redirect] = false
+      details[:redirect_to_external] = false
+      details[:redirect_to_https] = false
+      details[:live] = response.success?
+    end
+
+    details
+  end
+
+  def to_hash(http_only=false)
+    if http_only
+      {
+        :domain => domain.to_s,
+        :uri => uri.to_s,
+        :live => !!response,
+        :ssl => https?,
+        :enforce_https => enforce_https?,
+        :non_www => non_www?,
+        :redirect => redirect,
+        :headers => headers
+      }
```

Comment:
The above "http_only" mode from #22 is still here, but now inaccessible. It could be removed.